### ISSUE TITLE
How to remove spaces at the end of lines

### ISSUE THREAD

**Issue by ruv** [newren/git-filter-repo#145]
I wanted to remove spaces and `\r` at the end of lines  (i.e., also translate CRLF into LF line endings).

The following variant does not work:
`git filter-repo --replace-text <(echo 'regex:\s+$==>')` (1)

This one works:
`git filter-repo --replace-text <(echo 'regex:[\r\t ]+\n==>\n')` (2)

Although, it does not remove spaces at the end of a file, when the last line is not terminated with `\n`.

-----

The [documentation](https://github.com/newren/git-filter-repo/blob/7b3e714b94a6e5b9f478cb981c7f560ef3f36506/Documentation/git-filter-repo.txt#L875) says:
> globs and regexes are applied to each line of the file; it is not possible with --replace-text to match a multi-line string.

But it seems by the code, the regexp is applied to the full blob as is, [line 3250](https://github.com/newren/git-filter-repo/blob/7b3e714b94a6e5b9f478cb981c7f560ef3f36506/git-filter-repo#L3250):
```
blob.data = regex.sub(replacement, blob.data)
```

It explains why the variant (2) above works.

-----

So, what is the right way to remove spaces at the end of lines in accordance with the documentation?  (perhaps, the documentation and/or the code should be updated).

**Comment by newren** [comment]
Yeah, the documentation is wrong, sorry.

Anyway, the problem with the first variant is that '$' matches the end of the string, not the end of each line, as per python docs: https://docs.python.org/3/library/re.html.  You would need to force MULTILINE mode by injecting a `(?m)` into the regex (see the section of the documentation discussing `(?aiLmsux)` on python regular expressions at the same page linked above), so something like

git filter-repo --replace-text <(echo 'regex:(?m)\s+$==>')

should do the trick.


I'll keep this open to remind me to update the broken docs.  Thanks for pointing it out.

**Comment by newren** [comment]
Fixed in commit 9282a33a02 (git-filter-repo.txt: regexes & globs apply to entire file, not to lines, 2020-10-19).

Thanks for the report!

**Comment by ruv** [comment]
Just if somebody is interested: in the multiline mode `(?m)`, the pattern removes several consequential newlines. I tried:
```
regex:(?m)\s+$==>
regex:(?m)\s+?$==>
regex:(?m)\s+?$==>\n
regex:(?m)[\r\t ]+$==>\n
```
The reason is that both `\s` and `$` match `\n` in the multiline mode (in the single-line mode, `$` matches only the end of the input string/text).


The variant that works as it's required is the following one:
```
regex:[\r\t ]+(\n|$)==>\n
```

To convert both CRLF and CR to LF, and then remove trailing spaces:
```
regex:\r(\n)?==>\n'
regex:[ \t\v\f]+\n==>\n'
```
